rpc/server_interceptors.go

```diff
@@ -0,0 +1,126 @@
+			ctx, span = trace.StartSpanWithRemoteParent(ctx, "server_root", remoteSpanContext)
+			defer span.End()
+		} else {
+			logger.Warnf("client did not send valid Span metadata in headers, local Spans will not be linked to client. reason: %w", err)
+		}
+
+		resp, err := handler(ctx, req)
+		if err == nil {
+			return resp, nil
+		}
+		if _, ok := status.FromError(err); ok {
+			return nil, err
+		}
+		if s := status.FromContextError(err); s != nil {
+			return nil, s.Err()
+		}
+		return nil, err
+	}
+}
+
+// StreamServerTracingInterceptor starts a new Span if Span metadata exists in the context.
+func StreamServerTracingInterceptor(logger golog.Logger) grpc.StreamServerInterceptor {
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		remoteSpanContext, err := remoteSpanContextFromContext(stream.Context())
+		if err == nil {
+			newCtx, span := trace.StartSpanWithRemoteParent(stream.Context(), "server_root", remoteSpanContext)
+			defer span.End()
+			stream = wrapServerStream(newCtx, stream)
+		} else {
+			logger.Warnf("client did not send valid Span metadata in headers, local Spans will not be linked to client. reason: %w", err)
+		}
+
+		err = handler(srv, stream)
+		if err == nil {
+			return nil
+		}
+		if _, ok := status.FromError(err); ok {
+			return err
+		}
+		if s := status.FromContextError(err); s != nil {
+			return s.Err()
+		}
+		return err
+	}
+}
+
+type serverStreamWrapper struct {
+	grpc.ServerStream
+	ctx context.Context
+}
+
+func (s *serverStreamWrapper) Context() context.Context {
+	return s.ctx
+}
+
+func wrapServerStream(ctx context.Context, stream grpc.ServerStream) *serverStreamWrapper {
+	s := serverStreamWrapper{ServerStream: stream, ctx: ctx}
+	return &s
+}
+
+func remoteSpanContextFromContext(ctx context.Context) (trace.SpanContext, error) {
+	var err error
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return trace.SpanContext{}, errors.New("no metadata in context")
+	}
+
+	// Extract trace-id
+	traceIDMetadata := md.Get("trace-id")
+	if len(traceIDMetadata) == 0 {
+		return trace.SpanContext{}, errors.New("trace-id is missing from metadata")
+	}
+
+	traceIDBytes, err := hex.DecodeString(traceIDMetadata[0])
+	if err != nil {
+		return trace.SpanContext{}, fmt.Errorf("trace-id could not be decoded: %w", err)
+	}
+	var traceID trace.TraceID
+	copy(traceID[:], traceIDBytes)
+
+	// Extract span-id
+	spanIDMetadata := md.Get("span-id")
+	spanIDBytes, err := hex.DecodeString(spanIDMetadata[0])
+	if err != nil {
+		return trace.SpanContext{}, fmt.Errorf("span-id could not be decoded: %w", err)
+	}
+	var spanID trace.SpanID
+	copy(spanID[:], spanIDBytes)
+
+	// Extract trace-options
+	traceOptionsMetadata := md.Get("trace-options")
+	if len(traceOptionsMetadata) == 0 {
+		return trace.SpanContext{}, errors.New("trace-options is missing from metadata")
+	}
+
+	traceOptionsUint, err := strconv.ParseUint(traceOptionsMetadata[0], 10, 32)
```

Comment:
What these values (`10` and `32`) mean, or why they were chosen, was initially unclear to me, and I'd imagine an eng without context would also scratch their head about this. Could this be documented?